packages/backend/src/helpers/compute-parameters.ts

```diff
@@ -55,12 +56,30 @@ function findAndSubstituteVariables(
       if (isVariable) {
         const stepIdAndKeyPath = part.replace(/{{step.|}}/g, '') as string
         const [stepId, ...keyPaths] = stepIdAndKeyPath.split('.')
-        const keyPath = keyPaths.join('.')
         const executionStep = executionSteps.find((executionStep) => {
           return executionStep.stepId === stepId
         })
         const data = executionStep?.dataOut
-        const dataValue = get(data, keyPath)
+
+        const keyPath = keyPaths.join('.')
+        let dataValue = get(data, keyPath)
+        // custom logic to deal with backward compatibility of key encoding for
+        // data from vault. Under the new logic, data from vault will always have
+        // hex-encoded key while the old templates might still used non-encoded
+        // hence if the value is not defined and keysEncoded flag was set, we
+        // attempt to convert the template string to use hex-encoded key
+        if (
+          dataValue === undefined &&
+          executionStep.appKey === vaultWorkspace.key &&
+          data &&
+          data._metadata &&
+          (data._metadata as Record<string, any>).keysEncoded
```

Comment:
I think we can collapse the checks + add checks for nullibility and also keyPath length just in case. But most impt is to keep the vault workspace app check at the top so that everything short circuits and blast radius is contained to vault workspace

```suggestion
          dataValue === undefined &&
          executionStep?.appKey === vaultWorkspace.key &&
          (data?._metadata as IJSONObject)?.keysEncoded &&
          keyPaths.length > 0
```